raiden/api/python.py

```diff
@@ -242,29 +261,37 @@ def channel_deposit(
             raise InsufficientFunds(msg)
 
         netcontract_address = channel_state.identifier
-        token.approve(netcontract_address, amount)
-
         channel_proxy = self.raiden.chain.netting_channel(netcontract_address)
-        channel_proxy.deposit(amount)
 
-        old_balance = channel_state.our_state.contract_balance
-        target_balance = old_balance + amount
+        # If concurrent operations are happening on the channel, fail the request
+        if not channel_proxy.channel_operations_lock.acquire(0):
+            raise ChannelBusyError(
+                f"""Channel with id {channel_state.identifier} is
+                busy with another ongoing operation"""
+            )
 
-        msg = 'After {} seconds the deposit was not properly processed.'.format(
-            poll_timeout
-        )
+        with channel_proxy.channel_operations_lock:
+            token.approve(netcontract_address, amount)
```

Comment:
Humm, this is problematic. For this to work properly we always need to follow the above pattern (hold the channel proxy to call token's approve). This is too easy to mess up, ideally the exposed APIs should only support correct usage.

For now, I think we can just shorten the lock zone, and hold it inside the `channel_proxy` methods. This allows `token.approve` to race, which should be fine. In the future I believe the best option is to add a facade in the smart contract that calls both `approve` and `deposit`, allowing us to save a transaction and to execute it atomically.

Reply:
hey augusto, this is a re-entrant lock that's also being held by the channel_proxy methods. The changes are below in netting_channel.py